### PR TITLE
Forcing restart when toggling slow-mo

### PR DESCRIPTION
**Pull request by gushuro**
Fixing https://github.com/Microsoft/pxt-microbit/issues/1869.

- [x] cleaning up startSimulator/stopSimulator/restartSimulator to use the React state to determine if a debug build is necessary
- [x] use options bag in startSimulator to avoid unnoticed breaking changes
- [x] adding a "pending" state in the simulator driver that tells the IDE that an code operation has started (like blocks->js). pre-state before "starting" state.

### CODE REVIEW

File: webapp/src/app.tsx [hunk 1]
```diff
@@ -1864,15 +1864,16 @@ export class ProjectView
...
+        if (tracing) {
             this.editor.clearHighlightedStatements();
             simulator.setTraceInterval(0);
         }
         else {
             simulator.setTraceInterval(intervalSpeed != undefined ? intervalSpeed : simulator.SLOW_TRACE_INTERVAL);
         }
-        this.setState({ tracing: !this.state.tracing })
-        this.restartSimulator();
+        this.setState({ tracing: !tracing })
```
Comment (pelikhan) [inline webapp/src/app.tsx]: why is restartSimulator not hanlding the "tracing" mode?

File: webapp/src/app.tsx [hunk 2]
```diff
@@ -2020,6 +2021,15 @@ export class ProjectView
         this.blocksEditor.setBreakpointsFromBlocks();
     }
 
+    restartSimulatorTrace(trace?: boolean) {
+        if (this.state.simState == pxt.editor.SimState.Stopped
```
Comment (pelikhan) [inline webapp/src/app.tsx]: why is this not part of restartSimulator?

File: webapp/src/app.tsx [hunk 3]
```diff
@@ -1864,15 +1864,16 @@ export class ProjectView
...
 
     toggleTrace(intervalSpeed?: number) {
-        if (this.state.tracing) {
+        let tracing = this.state.tracing;
+        if (tracing) {
             this.editor.clearHighlightedStatements();
             simulator.setTraceInterval(0);
         }
         else {
             simulator.setTraceInterval(intervalSpeed != undefined ? intervalSpeed : simulator.SLOW_TRACE_INTERVAL);
```
Comment (pelikhan) [inline webapp/src/app.tsx]: ``!== undefined``

Comment (pelikhan) [reply webapp/src/app.tsx]: ACtualy you could just do ``intervalSpeed || simulator.SLOW_TRACE_INTERNAL``

File: webapp/src/app.tsx [hunk 4]
```diff
@@ -1864,15 +1864,16 @@ export class ProjectView
...
         }
-        this.setState({ tracing: !this.state.tracing })
-        this.restartSimulator();
+        this.setState({ tracing: !tracing })
+        this.startSimulator(true);
```
Comment (pelikhan) [inline webapp/src/app.tsx]: chain it in the setState -> setState({ tracing: !tracing}, () => this.startSimulator(true))

Comment (gushuro) [reply webapp/src/app.tsx]: Sure

File: webapp/src/app.tsx (hunk 3 above)
Comment (gushuro) [reply webapp/src/app.tsx]: Yep

File: webapp/src/app.tsx [hunk 5]
```diff
@@ -1864,15 +1864,15 @@ export class ProjectView
...
+        } else {
+            simulator.setTraceInterval(intervalSpeed || simulator.SLOW_TRACE_INTERVAL);
         }
-        else {
-            simulator.setTraceInterval(intervalSpeed != undefined ? intervalSpeed : simulator.SLOW_TRACE_INTERVAL);
-        }
-        this.setState({ tracing: !this.state.tracing })
-        this.restartSimulator();
+        this.setState({ tracing: !tracing }, () => this.startSimulator(true))
+        this.startSimulator(true);
```
Comment (pelikhan) [inline webapp/src/app.tsx]: you are starting twince the simulator + why is the argument true always?

File: webapp/src/app.tsx [hunk 6]
```diff
@@ -2144,16 +2154,15 @@ export class ProjectView
         pxt.log("turning debugging mode to " + state);
         this.setState({ debugging: state, tracing: false }, () => {
             this.renderCore()
+            const blocks = this.blocksEditor.editor.getAllBlocks();
```
Comment (pelikhan) [inline webapp/src/app.tsx]: it's not guaranteed that the blocks editor is loaded yet. You should not leak this abstraction here.

**Comment by pelikhan** [comment]
In genreal, app.tsx should have have access to the blockly workspace.

**Pull request merged**